src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerCopyFileFromContainer.groovy

```diff
@@ -15,55 +15,209 @@
 
-        getDestFile().withOutputStream  { OutputStream out ->
-            def input
+        if (!hostPath)
+            hostPath = "${project.projectDir}"
 
-            try {
-                input = dockerClient.copyFileFromContainerCmd(getContainerId(), getResource()).exec()
+        def containerCommand = dockerClient.copyFileFromContainerCmd(getContainerId(), getRemotePath())
+        logger.quiet "Copying '${getRemotePath()}' from container with ID '${getContainerId()}' to '${getHostPath()}'."
 
-                if (getCompressed()) {
-                    compressFile(out, input)
-                } else {
-                    out << input
-                }
-            }
-            finally {
-                input?.close()
+        def tarStream
+        try {
+
+            tarStream = containerCommand.exec()
+            def hostDestination = new File(hostPath)
+
+            // if compressed leave file as is otherwise untar
+            if (compressed) {
+                copyFileCompressed(tarStream, hostDestination)
+            } else {
+                copyFile(tarStream, hostDestination)
             }
+        } finally {
+            tarStream?.close()
         }
     }
 
-    private void compressFile(OutputStream outputStream, input) {
-        GZIPOutputStream gzipOut
+    /**
+     * Copy tar-stream from container to host
+     */
+    private void copyFileCompressed(InputStream tarStream, File hostDestination) {
 
-        try {
-            gzipOut = new GZIPOutputStream(outputStream)
-            gzipOut << input
+        // If user supplied an existing directory then we are responsible for naming and so
+        // will ensure file ends with '.tar'. If user supplied a regular file then use
+        // whichever name was passed in.
+        def fileName = new File(getRemotePath()).name
+        def compressedFileName = (hostDestination.exists() && hostDestination.isDirectory()) ?
+                (fileName.endsWith(".tar") ?: fileName + ".tar") :
+                hostDestination.name
+
+        def compressedFileLocation = (hostDestination.exists() && hostDestination.isDirectory()) ?
+                hostDestination :
+                hostDestination.parentFile
+
+        // If user supplied regular file ensure its parent location exists and if
+        // the regular file itself exists, delete to avoid clobbering.
+        if (hostDestination.exists()) {
+            if (!hostDestination.isDirectory())
+                hostDestination.delete()
+        } else {
+            hostDestination.parentFile.mkdirs()
         }
-        finally {
-            gzipOut?.close()
+
+        new File(compressedFileLocation,
+                compressedFileName).
+                withOutputStream { it << tarStream }
+
+    }
+
+    /**
+     * Copy regular file or directory from container to host
+     */
+    private void copyFile(InputStream tarStream, File hostDestination) {
```

Comment:
We'll need to break this up into methods. The method is more than 100 lines long and really hard to understand as one block of code. A good rule of thumb here is to break it up into methods where ever you added a comment.